packages/flutter_tools/gradle/src/main/groovy/shared.groovy

```diff
@@ -0,0 +1,56 @@
+import groovy.json.JsonSlurper
+
+class NativePluginLoader {
+    NativePluginLoader() {}
+
+    // This string must match _kFlutterPluginsHasNativeBuildKey defined in
+    // packages/flutter_tools/lib/src/flutter_plugins.dart.
+    static final String nativeBuildKey = 'native_build'
+    static final String flutterPluginsDependenciesFile = '.flutter-plugins-dependencies'
+
+    void forEachPlugin(File flutterSourceDirectory, Closure<Object> callback) {
+        def meta = getDependenciesMetadata(flutterSourceDirectory)
+        if (meta == null) {
+            return
+        }
+
+        assert meta.plugins instanceof Map
+        assert meta.plugins.android instanceof List
+        // Includes the Flutter plugins that support the Android platform.
+        meta.plugins.android.each { androidPlugin ->
+            assert androidPlugin.name instanceof String
+            assert androidPlugin.path instanceof String
+            // Skip plugins that have no native build (such as a Dart-only implementation
+            // of a federated plugin).
+            def needsBuild = androidPlugin.containsKey(nativeBuildKey) ? androidPlugin[nativeBuildKey] : true
+            if (!needsBuild) {
+                return
+            }
+            callback(androidPlugin)
+        }
+    }
+
+
+    private Map parsedFlutterPluginsDependencies
+
+    /**
+     * Parses <project-src>/.flutter-plugins-dependencies
+     */
+    Map getDependenciesMetadata(File flutterSourceDirectory) {
+        if (parsedFlutterPluginsDependencies) {
+            return parsedFlutterPluginsDependencies
+        }
+        File pluginsDependencyFile = new File(flutterSourceDirectory, flutterPluginsDependenciesFile)
+        if (pluginsDependencyFile.exists()) {
+            def object = new JsonSlurper().parseText(pluginsDependencyFile.text)
+            assert object instanceof Map
+            parsedFlutterPluginsDependencies = object
+            return object
+        }
+        return null
+    }
+}
+
+ext {
+    nativePluginLoader = new NativePluginLoader()
+}
```

Comment:
Can you do this in `flutter.groovy` to keep the logic in one place?

Reply:
> I can't access `NativePluginLoader` from within `flutter.groovy`

Hmm, why is that? These files are in the same directory, so should be in the same Java package.

Using `ext` to work around this seems like a hack to me.

Reply:
That was suspicious for me too.
If calling `NativePluginLoader().getPlugins(getFlutterSourceDirectory())` (even without import) it should work right?
But it doesn't: it gives:

```
A problem occurred evaluating script.
> Failed to apply plugin class 'FlutterPlugin'.
   > No signature of method: FlutterPlugin.NativePluginLoader() is applicable for argument types: () values: []
```
I don't know why. Have tried so many possibilities.

Reply:
I forgot the `new` keyword. Dummy me...

Reply:
But still cannot find the class:
```
Could not compile script 'D:\Tmp\Coding\git\flutter\flutter\packages\flutter_tools\gradle\src\main\groovy\flutter.groovy'.
> startup failed:
  script 'D:\Tmp\Coding\git\flutter\flutter\packages\flutter_tools\gradle\src\main\groovy\flutter.groovy': 629: unable to resolve class NativePluginLoader 
   @ line 629, column 16.
             return (new NativePluginLoader()).getPlugins(getFlutterSourceDirectory())
                    ^
```

Reply:
> Edit: maybe because only the flutter.groovy file is [applied](https://github.com/flutter/flutter/blob/e33d4b86270e3c012ba13d68d6e90f2eabc4912b/packages/flutter_tools/gradle/flutter.gradle#L9), and not the whole folder (?)

This makes sense. That `apply` inside `packages/flutter_tools/gradle/flutter.gradle` is used in projects that have not migrated the declarative apply of the Flutter Gradle Plugin using the `plugins {}` block (which, tbh, are most of existing projects - there was no announcement or anything). See #135392.

If your app project was `flutter create`d on latest `master`, then I think it'd work. Still, not good enough.

Reply:
So I leave it like it is right now (?)